From bff74736f0170d366fd10c0278cb55534355eec3 Mon Sep 17 00:00:00 2001
From: Shahed Shaikh <shahedx.shaikh@intel.com>
Date: Thu, 19 Mar 2020 14:16:47 +0800
Subject: [PATCH] DRV_DPM_SW-204: PRX improve sanity checks for exported APIs

- Add sanity check in exported APIs to check if DPM is initialized.
- Add new globle variable to mark CPU init is done.
- Check for valid inst and other arguments.

Signed-off-by: Shahed Shaikh <shahedx.shaikh@intel.com>
---
 drivers/net/datapath/dpm/datapath.h              |   1 +
 drivers/net/datapath/dpm/datapath_api.c          | 130 +++++++++++++++--
 drivers/net/datapath/dpm/datapath_misc.c         |  42 +++++-
 drivers/net/datapath/dpm/datapath_qos.c          | 178 ++++++++++++++++++++---
 drivers/net/datapath/dpm/datapath_swdev.c        |  10 ++
 drivers/net/datapath/dpm/datapath_tx.c           |   5 +
 drivers/net/datapath/dpm/gswip31/datapath_misc.c |   2 +
 7 files changed, 329 insertions(+), 39 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath.h b/drivers/net/datapath/dpm/datapath.h
index 1f7196ef094e..8f254310fa90 100644
--- a/drivers/net/datapath/dpm/datapath.h
+++ b/drivers/net/datapath/dpm/datapath.h
@@ -987,6 +987,7 @@ int dp_request_inst(struct dp_inst_info *info, u32 flag);
 int register_dp_cap(u32 flag);
 int bp_pmapper_dev_get(int inst, struct net_device *dev);
 extern int dp_init_ok;
+extern int dp_cpu_init_ok;
 void set_chksum(struct pmac_tx_hdr *pmac, u32 tcp_type,
 		u32 ip_offset, int ip_off_hw_adjust, u32 tcp_h_offset);
 
diff --git a/drivers/net/datapath/dpm/datapath_api.c b/drivers/net/datapath/dpm/datapath_api.c
index a57f7f119f83..383df1602ba7 100644
--- a/drivers/net/datapath/dpm/datapath_api.c
+++ b/drivers/net/datapath/dpm/datapath_api.c
@@ -50,6 +50,7 @@ char *dp_port_status_str[] = {
 };
 
 int dp_init_ok;
+int dp_cpu_init_ok;
 atomic_t dp_status = ATOMIC_INIT(0);
 DP_DEFINE_LOCK(dp_lock);
 u32 dp_dbg_err = 1; /*print error */
@@ -126,6 +127,11 @@ EXPORT_SYMBOL(dp_register_ops);
 
 void *dp_get_ops(int inst, enum DP_OPS_TYPE type)
 {
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return NULL;
+	}
+
 	if (is_invalid_inst(inst) || type >= DP_OPS_CNT) {
 		DP_DEBUG(DP_DBG_FLAG_REG, "wrong index\n");
 		return NULL;
@@ -206,6 +212,14 @@ int parser_enabled(int ep, struct dma_rx_desc_1 *desc_1)
 
 u32 *get_port_flag(int inst, int index)
 {
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return NULL;
+	}
+
+	if (is_invalid_inst(inst))
+		return NULL;
+
 	if (index < dp_port_prop[inst].info.cap.max_num_dp_ports)
 		return &get_dp_port_info(inst, index)->alloc_flags;
 
@@ -262,8 +276,14 @@ int8_t parser_size(int8_t v)
 int dp_set_gsw_parser(u8 flag, u8 cpu, u8 mpe1, u8 mpe2, u8 mpe3)
 {
 	int inst = 0;
-	struct inst_info *dp_info = get_dp_prop_info(inst);
+	struct inst_info *dp_info;
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	dp_info = get_dp_prop_info(inst);
 	if (!dp_info->dp_set_gsw_parser)
 		return -1;
 
@@ -274,8 +294,14 @@ EXPORT_SYMBOL(dp_set_gsw_parser);
 int dp_get_gsw_parser(u8 *cpu, u8 *mpe1, u8 *mpe2, u8 *mpe3)
 {
 	int inst = 0;
-	struct inst_info *dp_info = get_dp_prop_info(inst);
+	struct inst_info *dp_info;
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	dp_info = get_dp_prop_info(inst);
 	if (!dp_info->dp_get_gsw_parser)
 		return -1;
 
@@ -354,14 +380,14 @@ static int32_t dp_alloc_port_private(int inst,
 	int i;
 	struct cbm_dp_alloc_data cbm_data = {0};
 	struct pmac_port_info *port;
-	struct inst_info *dp_info = get_dp_prop_info(inst);
+	struct inst_info *dp_info;
 
 	if (!owner) {
 		pr_err("Allocate port failed for owner NULL\n");
 		return DP_FAILURE;
 	}
 
-	if (is_invalid_port(port_id))
+	if (is_invalid_port(port_id) || is_invalid_inst(inst))
 		return DP_FAILURE;
 
 #if IS_ENABLED(CONFIG_INTEL_DATAPATH_DBG)
@@ -375,6 +401,7 @@ static int32_t dp_alloc_port_private(int inst,
 	}
 #endif
 
+	dp_info = get_dp_prop_info(inst);
 	cbm_data.dp_inst = inst;
 	cbm_data.cbm_inst = dp_port_prop[inst].cbm_inst;
 	port = get_dp_port_info(inst, port_id);
@@ -978,6 +1005,11 @@ int32_t dp_update_subif_info(struct dp_subif_upd_info *info)
 	dp_get_netif_subifid_fn_t subifid_fn = NULL;
 	struct dp_subif_data old_data = {0}, new_data = {0};
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
 	if (!info || is_invalid_inst(info->inst) ||
 	    is_invalid_port(info->dp_port)) {
 		pr_err("dp_update_subif failed, info not valid\n");
@@ -1254,7 +1286,17 @@ EXPORT_SYMBOL(dp_alloc_port_ext);
 int32_t dp_register_dev(struct module *owner, u32 port_id,
 			dp_cb_t *dp_cb, u32 flags)
 {
-	int inst = dp_get_inst_via_module(owner, port_id, 0);
+	int inst;
+
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!owner || is_invalid_port(port_id))
+		return DP_FAILURE;
+
+	inst = dp_get_inst_via_module(owner, port_id, 0);
 
 	if (inst < 0) {
 		pr_err("%s not valid module %s\n", __func__, owner->name);
@@ -1294,6 +1336,7 @@ int32_t dp_register_dev_ext(int inst, struct module *owner, u32 port_id,
 		pr_err("dp_register_dev failed for datapath not init yet\n");
 		return DP_FAILURE;
 	}
+
 	if (!data)
 		data = &tmp_data;
 
@@ -1420,6 +1463,13 @@ int32_t dp_register_subif_ext(int inst, struct module *owner,
 			 "dp_register_subif fail for datapath not init yet\n");
 		return DP_FAILURE;
 	}
+
+	if (!subif_id || !owner) {
+		DP_DEBUG(DP_DBG_FLAG_REG, "%s Failed subif_id %px owner %px\n",
+			 __func__, subif_id, owner);
+		return DP_FAILURE;
+	}
+
 	DP_DEBUG(DP_DBG_FLAG_REG,
 		 "%s:owner=%s dev=%s %s=%s port_id=%d subif=%d(%s) %s%s\n",
 		 (flags & DP_F_DEREGISTER) ?
@@ -1435,12 +1485,6 @@ int32_t dp_register_subif_ext(int inst, struct module *owner,
 		 (flags & DP_F_ALLOC_EXPLICIT_SUBIFID) ?
 		 "Explicit" : "Non-Explicit");
 
-	if (!subif_id || !owner) {
-		DP_DEBUG(DP_DBG_FLAG_REG, "%s Failed subif_id %px owner %px\n",
-			 __func__, subif_id, owner);
-		return DP_FAILURE;
-	}
-
 	if (is_invalid_port(subif_id->port_id) || is_invalid_inst(inst))
 		return DP_FAILURE;
 
@@ -1497,6 +1541,11 @@ int32_t dp_register_subif(struct module *owner, struct net_device *dev,
 	int inst;
 	struct dp_subif_data data = {0};
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
 	if (!subif_id || !owner || is_invalid_port(subif_id->port_id)) {
 		DP_DEBUG(DP_DBG_FLAG_REG, "%s fail owner %px subif_id %px\n",
 			 __func__, subif_id, owner);
@@ -1524,6 +1573,11 @@ int32_t dp_get_netif_subifid(struct net_device *netif, struct sk_buff *skb,
 	dp_get_netif_subifid_fn_t subifid_fn_t;
 	int res = DP_FAILURE;
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return res;
+	}
+
 	idx = dp_subif_hash(netif);
 	//TODO handle DSL case in future
 	rcu_read_lock_bh();
@@ -1876,6 +1930,11 @@ int dp_get_port_subitf_via_dev(struct net_device *dev, dp_subif_t *subif)
 {
 	int res;
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
 	DP_LIB_LOCK(&dp_lock);
 	res = dp_get_port_subitf_via_dev_private(dev, subif);
 	DP_LIB_UNLOCK(&dp_lock);
@@ -1917,6 +1976,11 @@ int dp_get_port_subitf_via_ifname(char *ifname, dp_subif_t *subif)
 	int res;
 	struct net_device *dev;
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
 	if (!ifname)
 		return -1;
 	dev = dev_get_by_name(&init_net, ifname);
@@ -2092,6 +2156,11 @@ int dp_get_netif_stats(struct net_device *dev, dp_subif_t *subif_id,
 		       struct rtnl_link_stats64 * stats,
 		       u32 flags);
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
 	if (subif_id) {
 		subif = *subif_id;
 	} else if (dev) {
@@ -2122,6 +2191,11 @@ int dp_clear_netif_stats(struct net_device *dev, dp_subif_t *subif_id,
 	int (*clear_netif_mib_fn)(dp_subif_t *subif, void *priv, u32 flag);
 	int i;
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
 	if (subif_id) {
 		clear_netif_mib_fn =
 			get_dp_prop_info(subif_id->inst)->dp_clear_netif_mib;
@@ -2156,6 +2230,11 @@ int dp_pmac_set(int inst, u32 port, dp_pmac_cfg_t *pmac_cfg)
 {
 	int (*dp_pmac_set_fn)(int inst, u32 port, dp_pmac_cfg_t *pmac_cfg);
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
 	if (is_invalid_inst(inst))
 		return DP_FAILURE;
 
@@ -2432,9 +2511,12 @@ u32 get_dp_dbg_flag(void)
  */
 int dp_get_cap(struct dp_cap *cap, int flag)
 {
-	if (!cap)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
 		return DP_FAILURE;
-	if (is_invalid_inst(cap->inst))
+	}
+
+	if (!cap || is_invalid_inst(cap->inst))
 		return DP_FAILURE;
 	if (!hw_cap_list[cap->inst].valid)
 		return DP_FAILURE;
@@ -2459,6 +2541,11 @@ int dp_rx_enable(struct net_device *netif, char *ifname, u32 flags)
 	struct pmac_port_info *port_info;
 	int vap;
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
 	if (dp_get_netif_subifid(netif, NULL, NULL, NULL, &subif, 0)) {
 		DP_DEBUG(DP_DBG_FLAG_DBG, "get subifid fail(%s)\n",
 			 netif ? netif->name : "NULL");
@@ -2521,6 +2608,11 @@ int dp_set_bp_attr(struct dp_bp_attr *conf, u32 flag)
 	int ret = DP_SUCCESS;
 	struct inst_info *dp_info;
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
 	if (!conf) {
 		DP_DEBUG(DP_DBG_FLAG_DBG, "conf passed is (%s)\n",
 			 conf ? conf->dev->name : "NULL");
@@ -2625,7 +2717,7 @@ int dp_register_event_cb(struct dp_event *info, u32 flag)
 {
 	int ret;
 
-	if (is_invalid_inst(info->inst))
+	if (!info || is_invalid_inst(info->inst))
 		return DP_FAILURE;
 
 	if (flag & DP_F_DEREGISTER)
@@ -2642,6 +2734,11 @@ int dp_free_buffer_by_policy(struct dp_buffer_info *info, u32 flag)
 	struct cqm_bm_free data = {0};
 	int ret;
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
 	if (!info)
 		return DP_FAILURE;
 
@@ -2748,6 +2845,11 @@ void dp_cleanup_module(void)
 /* This API will be enhanced later based on netdevice */
 int dp_get_mtu_size(struct net_device *dev, u32 *mtu_size)
 {
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
 	return cbm_get_mtu_size(mtu_size);
 }
 EXPORT_SYMBOL(dp_get_mtu_size);
diff --git a/drivers/net/datapath/dpm/datapath_misc.c b/drivers/net/datapath/dpm/datapath_misc.c
index 5e0a8b9cf588..71a9ffd380d6 100644
--- a/drivers/net/datapath/dpm/datapath_misc.c
+++ b/drivers/net/datapath/dpm/datapath_misc.c
@@ -726,6 +726,16 @@ int dp_ingress_ctp_tc_map_set(struct dp_tc_cfg *tc, int flag)
 	struct dp_meter_subif mtr_subif = {0};
 	struct inst_info *dp_info;
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!tc) {
+		pr_err("%s: tc_cfg is NULL\n", __func__);
+		return DP_FAILURE;
+	}
+
 	if (dp_get_netif_subifid(tc->dev, NULL, NULL, NULL,
 				 &mtr_subif.subif, 0)) {
 		DP_DEBUG(DP_DBG_FLAG_DBG, "get subifid fail(%s)\n",
@@ -743,7 +753,17 @@ EXPORT_SYMBOL(dp_ingress_ctp_tc_map_set);
 
 int dp_meter_alloc(int inst, int *meterid, int flag)
 {
-	struct inst_info *dp_info = get_dp_prop_info(inst);
+	struct inst_info *dp_info;
+
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!meterid || is_invalid_inst(inst))
+		return DP_FAILURE;
+
+	dp_info = get_dp_prop_info(inst);
 
 	if (!dp_info->dp_meter_alloc)
 		return DP_FAILURE;
@@ -757,6 +777,16 @@ int dp_meter_add(struct net_device *dev, struct dp_meter_cfg *meter, int flag)
 	struct dp_meter_subif mtr_subif = {0};
 	struct inst_info *dp_info;
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!dev || !meter) {
+		pr_err("%s failed: dev or meter_cfg can not be NULL\n", __func__);
+		return DP_FAILURE;
+	}
+
 	if ((flag & DP_METER_ATTACH_CTP) ||
 	    (flag & DP_METER_ATTACH_BRPORT) ||
 	    (flag & DP_METER_ATTACH_PCE)) {
@@ -792,6 +822,16 @@ int dp_meter_del(struct net_device *dev, struct dp_meter_cfg *meter, int flag)
 	struct dp_meter_subif mtr_subif = {0};
 	struct inst_info *dp_info;
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!dev || !meter) {
+		pr_err("%s failed: dev or meter_cfg can not be NULL\n", __func__);
+		return DP_FAILURE;
+	}
+
 	if ((flag & DP_METER_ATTACH_CTP) ||
 	    (flag & DP_METER_ATTACH_BRPORT) ||
 	    (flag & DP_METER_ATTACH_PCE)) {
diff --git a/drivers/net/datapath/dpm/datapath_qos.c b/drivers/net/datapath/dpm/datapath_qos.c
index 6fae998a4c36..ca8ac2f6ac32 100644
--- a/drivers/net/datapath/dpm/datapath_qos.c
+++ b/drivers/net/datapath/dpm/datapath_qos.c
@@ -11,7 +11,12 @@ int dp_node_link_add(struct dp_node_link *info, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!info)
+	if (unlikely(!dp_init_ok && !dp_cpu_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!info || is_invalid_inst(info->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(info->inst);
@@ -26,7 +31,12 @@ int dp_node_unlink(struct dp_node_link *info, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!info)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!info || is_invalid_inst(info->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(info->inst);
@@ -41,7 +51,12 @@ int dp_node_link_get(struct dp_node_link *info, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!info)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!info || is_invalid_inst(info->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(info->inst);
@@ -56,7 +71,12 @@ int dp_node_link_en_set(struct dp_node_link_enable *en, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!en)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!en || is_invalid_inst(en->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(en->inst);
@@ -71,7 +91,12 @@ int dp_node_link_en_get(struct dp_node_link_enable *en, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!en)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!en || is_invalid_inst(en->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(en->inst);
@@ -86,7 +111,12 @@ int dp_link_add(struct dp_qos_link *cfg, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!cfg)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!cfg || is_invalid_inst(cfg->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(cfg->inst);
@@ -101,7 +131,12 @@ int dp_link_get(struct dp_qos_link *cfg, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!cfg)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!cfg || is_invalid_inst(cfg->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(cfg->inst);
@@ -116,7 +151,12 @@ int dp_qos_link_prio_set(struct dp_node_prio *info, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!info)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!info || is_invalid_inst(info->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(info->inst);
@@ -131,7 +171,12 @@ int dp_qos_link_prio_get(struct dp_node_prio *info, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!info)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!info || is_invalid_inst(info->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(info->inst);
@@ -146,7 +191,12 @@ int dp_queue_conf_set(struct dp_queue_conf *cfg, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!cfg)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!cfg || is_invalid_inst(cfg->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(cfg->inst);
@@ -161,6 +211,11 @@ int dp_queue_conf_get(struct dp_queue_conf *cfg, int flag)
 {
 	struct inst_info *dp_info;
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
 	if (!cfg)
 		return DP_FAILURE;
 
@@ -176,6 +231,11 @@ int dp_shaper_conf_set(struct dp_shaper_conf *cfg, int flag)
 {
 	struct inst_info *dp_info;
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
 	if (!cfg)
 		return DP_FAILURE;
 
@@ -191,7 +251,12 @@ int dp_shaper_conf_get(struct dp_shaper_conf *cfg, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!cfg)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!cfg || is_invalid_inst(cfg->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(cfg->inst);
@@ -206,7 +271,12 @@ int dp_node_alloc(struct dp_node_alloc *node, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!node)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!node || is_invalid_inst(node->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(node->inst);
@@ -221,7 +291,12 @@ int dp_node_free(struct dp_node_alloc *node, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!node)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!node || is_invalid_inst(node->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(node->inst);
@@ -236,7 +311,12 @@ int dp_node_children_free(struct dp_node_alloc *node, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!node)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!node || is_invalid_inst(node->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(node->inst);
@@ -253,7 +333,12 @@ int dp_deq_port_res_get(struct dp_dequeue_res *res, int flag)
 	struct pmac_port_info *port_info;
 	struct inst_info *dp_info;
 
-	if (!res)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!res || is_invalid_inst(res->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(res->inst);
@@ -285,7 +370,12 @@ int dp_counter_mode_set(struct dp_counter_conf *cfg, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!cfg)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!cfg || is_invalid_inst(cfg->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(cfg->inst);
@@ -300,7 +390,12 @@ int dp_counter_mode_get(struct dp_counter_conf *cfg, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!cfg)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!cfg || is_invalid_inst(cfg->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(cfg->inst);
@@ -315,7 +410,12 @@ int dp_queue_map_set(struct dp_queue_map_set *cfg, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!cfg)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!cfg || is_invalid_inst(cfg->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(cfg->inst);
@@ -330,7 +430,12 @@ int dp_queue_map_get(struct dp_queue_map_get *cfg, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!cfg)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!cfg || is_invalid_inst(cfg->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(cfg->inst);
@@ -345,7 +450,12 @@ int dp_children_get(struct dp_node_child *cfg, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!cfg)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!cfg || is_invalid_inst(cfg->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(cfg->inst);
@@ -360,7 +470,12 @@ int dp_qos_level_get(struct dp_qos_level *cfg, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!cfg)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!cfg || is_invalid_inst(cfg->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(cfg->inst);
@@ -375,7 +490,12 @@ int dp_qos_get_q_logic(struct dp_qos_q_logic *cfg, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!cfg)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!cfg || is_invalid_inst(cfg->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(cfg->inst);
@@ -392,7 +512,12 @@ int dp_qos_global_info_get(struct dp_qos_cfg_info *info, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!info)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!info || is_invalid_inst(info->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(info->inst);
@@ -407,7 +532,12 @@ int dp_qos_port_conf_set(struct dp_port_cfg_info *info, int flag)
 {
 	struct inst_info *dp_info;
 
-	if (!info)
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!info || is_invalid_inst(info->inst))
 		return DP_FAILURE;
 
 	dp_info = get_dp_prop_info(info->inst);
diff --git a/drivers/net/datapath/dpm/datapath_swdev.c b/drivers/net/datapath/dpm/datapath_swdev.c
index 5fc3e618ca7f..86a7d8d0d408 100644
--- a/drivers/net/datapath/dpm/datapath_swdev.c
+++ b/drivers/net/datapath/dpm/datapath_swdev.c
@@ -66,6 +66,16 @@ int dp_get_fid_by_brname(struct net_device *dev, int *inst)
 {
 	struct br_info *br_info;
 
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
+	if (!dev) {
+		pr_err("%s failed: dev can not be NULL\n", __func__);
+		return DP_FAILURE;
+	}
+
 	br_info = dp_swdev_bridge_entry_lookup(dev->name);
 	if (!br_info)
 		return -1;
diff --git a/drivers/net/datapath/dpm/datapath_tx.c b/drivers/net/datapath/dpm/datapath_tx.c
index 83bc6908a6b1..b4b2c8adeb01 100644
--- a/drivers/net/datapath/dpm/datapath_tx.c
+++ b/drivers/net/datapath/dpm/datapath_tx.c
@@ -298,6 +298,11 @@ int dp_tx_init(int inst)
 
 int dp_register_tx(enum DP_TX_PRIORITY priority, tx_fn fn, void *priv)
 {
+	if (unlikely(!dp_init_ok)) {
+		pr_err("%s failed: datapath not initialized yet\n", __func__);
+		return DP_FAILURE;
+	}
+
 	if (!dp_tx_ctx)
 		return -ENOMEM;
 	dp_tx_register_process(priority, fn, priv);
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_misc.c b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
index db77f6e022ed..4b47068597c5 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_misc.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
@@ -1053,6 +1053,8 @@ int dp_platform_queue_set(int inst, u32 flag)
 	port_info->deq_port_base = 0;
 	port_info->deq_port_num = 4;  /*need improve later*/
 
+	dp_cpu_init_ok = 1;
+
 	for (i = 0; i < ARRAY_SIZE(cpu_data.dq_tx_push_info); i++) {
 		struct dp_subif_info *sif;
 
